# Customer Intelligence System

## 1. Business Problem
The company wants to reduce customer churn and better understand customer segments. 
Churn leads to revenue loss. We aim to identify high-risk customers and group customers based on behavior.

## 2. ML Formulation
A) Churn Prediction → Binary Classification  
B) Customer Segmentation → Clustering

## 3. Target Variable
Churn Value(1/0)

## 4. Success Metrics
Classification: F1-score, ROC-AUC  
Clustering: Silhouette Score

## 5. Final Output
Churn probability + Risk level + Segment label

---

# About Dataset
## Context
A fictional telco company that provided home phone and Internet services to 7043 customers in California in Q3.

## Data Description
7043 observations with 33 variables

**CustomerID**: A unique ID that identifies each customer.

**Count**: A value used in reporting/dashboarding to sum up the number of customers in a filtered set.

**Country**: The country of the customer’s primary residence.

**State**: The state of the customer’s primary residence.

**City**: The city of the customer’s primary residence.

**Zip Code**: The zip code of the customer’s primary residence.

**Lat Long**: The combined latitude and longitude of the customer’s primary residence.

**Latitude**: The latitude of the customer’s primary residence.

**Longitude**: The longitude of the customer’s primary residence.

**Gender**: The customer’s gender: Male, Female

**Senior Citizen**: Indicates if the customer is 65 or older: Yes, No

**Partner**: Indicate if the customer has a partner: Yes, No

**Dependents**: Indicates if the customer lives with any dependents: Yes, No. Dependents could be children, parents, grandparents, etc.

**Tenure Months**: Indicates the total amount of months that the customer has been with the company by the end of the quarter specified above.

**Phone Service**: Indicates if the customer subscribes to home phone service with the company: Yes, No

**Multiple Lines**: Indicates if the customer subscribes to multiple telephone lines with the company: Yes, No

**Internet Service**: Indicates if the customer subscribes to Internet service with the company: No, DSL, Fiber Optic, Cable.

**Online Security**: Indicates if the customer subscribes to an additional online security service provided by the company: Yes, No

**Online Backup**: Indicates if the customer subscribes to an additional online backup service provided by the company: Yes, No

**Device Protection**: Indicates if the customer subscribes to an additional device protection plan for their Internet equipment provided by the company: Yes, No

**Tech Support**: Indicates if the customer subscribes to an additional technical support plan from the company with reduced wait times: Yes, No

**Streaming TV**: Indicates if the customer uses their Internet service to stream television programing from a third party provider: Yes, No. The company does not charge an additional fee for this service.

**Streaming Movies**: Indicates if the customer uses their Internet service to stream movies from a third party provider: Yes, No. The company does not charge an additional fee for this service.

**Contract**: Indicates the customer’s current contract type: Month-to-Month, One Year, Two Year.

**Paperless Billing**: Indicates if the customer has chosen paperless billing: Yes, No

**Payment Method**: Indicates how the customer pays their bill: Bank Withdrawal, Credit Card, Mailed Check

**Monthly Charge**: Indicates the customer’s current total monthly charge for all their services from the company.

**Total Charges**: Indicates the customer’s total charges, calculated to the end of the quarter specified above.

**Churn Label**: Yes = the customer left the company this quarter. No = the customer remained with the company. Directly related to Churn Value.

**Churn Value**: 1 = the customer left the company this quarter. 0 = the customer remained with the company. Directly related to Churn Label.

**Churn Score**: A value from 0-100 that is calculated using the predictive tool IBM SPSS Modeler. The model incorporates multiple factors known to cause churn. The higher the score, the more likely the customer will churn.

**CLTV**: Customer Lifetime Value. A predicted CLTV is calculated using corporate formulas and existing data. The higher the value, the more valuable the customer. High value customers should be monitored for churn.

**Churn Reason**: A customer’s specific reason for leaving the company. Directly related to Churn Category.

---

# Importing Libraries

In [82]:
import pandas as pd
import matplotlib.pyplot as plt
import numpy as np
%matplotlib inline

# 1. Data Acquisition

In [83]:
df = pd.read_excel('./Data/Telco_customer_churn.xlsx')
df.head()

CustomerID  Count        Country       State         City  Zip Code  \
0  3668-QPYBK      1  United States  California  Los Angeles     90003   
1  9237-HQITU      1  United States  California  Los Angeles     90005   
2  9305-CDSKC      1  United States  California  Los Angeles     90006   
3  7892-POOKP      1  United States  California  Los Angeles     90010   
4  0280-XJGEX      1  United States  California  Los Angeles     90015   

                 Lat Long   Latitude   Longitude  Gender  ...        Contract  \
0  33.964131, -118.272783  33.964131 -118.272783    Male  ...  Month-to-month   
1   34.059281, -118.30742  34.059281 -118.307420  Female  ...  Month-to-month   
2  34.048013, -118.293953  34.048013 -118.293953  Female  ...  Month-to-month   
3  34.062125, -118.315709  34.062125 -118.315709  Female  ...  Month-to-month   
4  34.039224, -118.266293  34.039224 -118.266293    Male  ...  Month-to-month   

  Paperless Billing             Payment Method  Monthly Charges Total Charges  \
0               Yes               Mailed check            53.85        108.15   
1               Yes           Electronic check            70.70        151.65   
2               Yes           Electronic check            99.65         820.5   
3               Yes           Electronic check           104.80       3046.05   
4               Yes  Bank transfer (automatic)           103.70        5036.3   

  Churn Label Churn Value Churn Score  CLTV                   Churn Reason  
0         Yes           1          86  3239   Competitor made better offer  
1         Yes           1          67  2701                          Moved  
2         Yes           1          86  5372                          Moved  
3         Yes           1          84  5003                          Moved  
4         Yes           1          89  5340  Competitor had better devices  

[5 rows x 33 columns]

In [84]:
# Assistant
import pandas as pd
import matplotlib.pyplot as plt
%matplotlib inline  

In [85]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7043 entries, 0 to 7042
Data columns (total 33 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   CustomerID         7043 non-null   object 
 1   Count              7043 non-null   int64  
 2   Country            7043 non-null   object 
 3   State              7043 non-null   object 
 4   City               7043 non-null   object 
 5   Zip Code           7043 non-null   int64  
 6   Lat Long           7043 non-null   object 
 7   Latitude           7043 non-null   float64
 8   Longitude          7043 non-null   float64
 9   Gender             7043 non-null   object 
 10  Senior Citizen     7043 non-null   object 
 11  Partner            7043 non-null   object 
 12  Dependents         7043 non-null   object 
 13  Tenure Months      7043 non-null   int64  
 14  Phone Service      7043 non-null   object 
 15  Multiple Lines     7043 non-null   object 
 16  Internet Service   7043 

`Total Charges` should be of type float so we will change it

In [86]:
df[df['Total Charges'] == ' '].count()

CustomerID           11
Count                11
Country              11
State                11
City                 11
Zip Code             11
Lat Long             11
Latitude             11
Longitude            11
Gender               11
Senior Citizen       11
Partner              11
Dependents           11
Tenure Months        11
Phone Service        11
Multiple Lines       11
Internet Service     11
Online Security      11
Online Backup        11
Device Protection    11
Tech Support         11
Streaming TV         11
Streaming Movies     11
Contract             11
Paperless Billing    11
Payment Method       11
Monthly Charges      11
Total Charges        11
Churn Label          11
Churn Value          11
Churn Score          11
CLTV                 11
Churn Reason          0
dtype: int64

We see that we have missing values in the `Total Charges` column

In [87]:
df = df.replace({'Total Charges': {' ': np.nan}})

C:\Users\HP\AppData\Local\Temp\ipykernel_2724\1081335959.py:1: FutureWarning: Downcasting behavior in `replace` is deprecated and will be removed in a future version. To retain the old behavior, explicitly call `result.infer_objects(copy=False)`. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`
  df = df.replace({'Total Charges': {' ': np.nan}})


In [88]:
df.describe()

Count      Zip Code     Latitude    Longitude  Tenure Months  \
count  7043.0   7043.000000  7043.000000  7043.000000    7043.000000   
mean      1.0  93521.964646    36.282441  -119.798880      32.371149   
std       0.0   1865.794555     2.455723     2.157889      24.559481   
min       1.0  90001.000000    32.555828  -124.301372       0.000000   
25%       1.0  92102.000000    34.030915  -121.815412       9.000000   
50%       1.0  93552.000000    36.391777  -119.730885      29.000000   
75%       1.0  95351.000000    38.224869  -118.043237      55.000000   
max       1.0  96161.000000    41.962127  -114.192901      72.000000   

       Monthly Charges  Total Charges  Churn Value  Churn Score         CLTV  
count      7043.000000    7032.000000  7043.000000  7043.000000  7043.000000  
mean         64.761692    2283.300441     0.265370    58.699418  4400.295755  
std          30.090047    2266.771362     0.441561    21.525131  1183.057152  
min          18.250000      18.800000     0.000000     5.000000  2003.000000  
25%          35.500000     401.450000     0.000000    40.000000  3469.000000  
50%          70.350000    1397.475000     0.000000    61.000000  4527.000000  
75%          89.850000    3794.737500     1.000000    75.000000  5380.500000  
max         118.750000    8684.800000     1.000000   100.000000  6500.000000

We see that the only feature with missing values is the `Churn Reason`

In [98]:
df.describe(include=['object'])

CustomerID        Country       State         City  \
count         7043           7043        7043         7043   
unique        7043              1           1         1129   
top     3668-QPYBK  United States  California  Los Angeles   
freq             1           7043        7043          305   

                      Lat Long Gender Senior Citizen Partner Dependents  \
count                     7043   7043           7043    7043       7043   
unique                    1652      2              2       2          2   
top     33.964131, -118.272783   Male             No      No         No   
freq                         5   3555           5901    3641       5416   

       Phone Service  ... Online Backup Device Protection Tech Support  \
count           7043  ...          7043              7043         7043   
unique             2  ...             3                 3            3   
top              Yes  ...            No                No           No   
freq            6361  ...          3088              3095         3473   

       Streaming TV Streaming Movies        Contract Paperless Billing  \
count          7043             7043            7043              7043   
unique            3                3               3                 2   
top              No               No  Month-to-month               Yes   
freq           2810             2785            3875              4171   

          Payment Method Churn Label                Churn Reason  
count               7043        7043                        1869  
unique                 4           2                          20  
top     Electronic check          No  Attitude of support person  
freq                2365        5174                         192  

[4 rows x 23 columns]

### Target Distribution

In [89]:
counts = df['Churn Value'].value_counts()

total = counts[0] + counts[1]
percentage_0 = (counts[0] / total)*100
percentage_1 = (counts[1] / total)*100

print(f"Percentage of Records with Churn Value 0: {np.round(percentage_0, 2)}%")
print(f"Percentage of Records with Churn Value 1: {np.round(percentage_1, 2)}%")

Percentage of Records with Churn Value 0: 73.46%
Percentage of Records with Churn Value 1: 26.54%


We see that we have only 26.54% of the records indicating customers that churned. Making this dataset imbalanced.

---

# 2. EDA

First off we remove some features that will not provide us any value if we keep them, like `CustomerID` and `Count`. And we also remove features that will only mean something if a cutomer churns, and have no value for us beforehand. These are `Churn Label`, `Churn Score`, and `Churn Reason`. We will also drop the `Lat Long` column as we already have the value of it as separate columns. We also see that `Country` and `State` have only one value, so we don't need them either.

In [99]:
new_df = df.drop(['CustomerID', 'Count', 'Country', 'State', 'Lat Long', 'Churn Label', 'Churn Score', 'Churn Reason'], axis=1)

In [100]:
new_df.head()

City  Zip Code   Latitude   Longitude  Gender Senior Citizen  \
0  Los Angeles     90003  33.964131 -118.272783    Male             No   
1  Los Angeles     90005  34.059281 -118.307420  Female             No   
2  Los Angeles     90006  34.048013 -118.293953  Female             No   
3  Los Angeles     90010  34.062125 -118.315709  Female             No   
4  Los Angeles     90015  34.039224 -118.266293    Male             No   

  Partner Dependents  Tenure Months Phone Service  ... Tech Support  \
0      No         No              2           Yes  ...           No   
1      No        Yes              2           Yes  ...           No   
2      No        Yes              8           Yes  ...           No   
3     Yes        Yes             28           Yes  ...          Yes   
4      No        Yes             49           Yes  ...           No   

  Streaming TV Streaming Movies        Contract Paperless Billing  \
0           No               No  Month-to-month               Yes   
1           No               No  Month-to-month               Yes   
2          Yes              Yes  Month-to-month               Yes   
3          Yes              Yes  Month-to-month               Yes   
4          Yes              Yes  Month-to-month               Yes   

              Payment Method Monthly Charges Total Charges Churn Value  CLTV  
0               Mailed check           53.85        108.15           1  3239  
1           Electronic check           70.70        151.65           1  2701  
2           Electronic check           99.65        820.50           1  5372  
3           Electronic check          104.80       3046.05           1  5003  
4  Bank transfer (automatic)          103.70       5036.30           1  5340  

[5 rows x 25 columns]

Now let us check for null values in our dataset

In [101]:
new_df.isnull().sum()

City                  0
Zip Code              0
Latitude              0
Longitude             0
Gender                0
Senior Citizen        0
Partner               0
Dependents            0
Tenure Months         0
Phone Service         0
Multiple Lines        0
Internet Service      0
Online Security       0
Online Backup         0
Device Protection     0
Tech Support          0
Streaming TV          0
Streaming Movies      0
Contract              0
Paperless Billing     0
Payment Method        0
Monthly Charges       0
Total Charges        11
Churn Value           0
CLTV                  0
dtype: int64

In [102]:
new_df.duplicated().sum()

np.int64(0)

We see that we have null values in `Total Charges`, and no duplicate records

In [103]:
new_df = new_df.fillna({'Total Charges': {np.nan : df['Total Charges'].mean()}})

We replace the null value with the mean of the column

## Correlation Analysis

In [104]:
new_df.columns

Index(['City', 'Zip Code', 'Latitude', 'Longitude', 'Gender', 'Senior Citizen',
       'Partner', 'Dependents', 'Tenure Months', 'Phone Service',
       'Multiple Lines', 'Internet Service', 'Online Security',
       'Online Backup', 'Device Protection', 'Tech Support', 'Streaming TV',
       'Streaming Movies', 'Contract', 'Paperless Billing', 'Payment Method',
       'Monthly Charges', 'Total Charges', 'Churn Value', 'CLTV'],
      dtype='object')

So save time, let us first plot the categorical features against our categorical target

In [106]:
for column in new_df.columns:
    print(f"Unique values in {column}: {new_df[column].unique()}")

Unique values in City: ['Los Angeles' 'Beverly Hills' 'Huntington Park' ... 'Standish' 'Tulelake'
 'Olympic Valley']
Unique values in Zip Code: [90003 90005 90006 ... 96128 96134 96146]
Unique values in Latitude: [33.964131 34.059281 34.048013 ... 40.346634 41.813521 39.191797]
Unique values in Longitude: [-118.272783 -118.30742  -118.293953 ... -120.386422 -121.492666
 -120.212401]
Unique values in Gender: ['Male' 'Female']
Unique values in Senior Citizen: ['No' 'Yes']
Unique values in Partner: ['No' 'Yes']
Unique values in Dependents: ['No' 'Yes']
Unique values in Tenure Months: [ 2  8 28 49 10  1 47 17  5 34 11 15 18  9  7 12 25 68 55 37  3 27 20  4
 58 53 13  6 19 59 16 52 24 32 38 54 43 63 21 69 22 61 60 48 40 23 39 35
 56 65 33 30 45 46 62 70 50 44 71 26 14 41 66 64 29 42 67 51 31 57 36 72
  0]
Unique values in Phone Service: ['Yes' 'No']
Unique values in Multiple Lines: ['No' 'Yes' 'No phone service']
Unique values in Internet Service: ['DSL' 'Fiber optic' 'No']
Unique values in

For our categorical features, we see that we have values `Yes`, `No`, `No internet service`, and `No phone service`. We will map these to 1 and 0, `1 for Yes`, and `0 for No, No internet service, No phone service`

In [118]:
for column in new_df.columns:
    unique_vals = set(new_df[column].dropna().unique())
    if unique_vals.issubset({'Yes', 'No', 'No internet service', 'No phone service'}):
        # Fix: Apply replace directly to the column, not the entire dataframe
        new_df[column] = new_df[column].replace({
            'Yes': 1,
            'No': 0,
            'No internet service': 0,
            'No phone service': 0
        })

Our one more categorical feature that can be replaced with 1/0 is the gender, we will map `1 for Male` and `0 for Female`

In [119]:
new_df['Gender'] = new_df['Gender'].replace({
    'Male': 1,
    'Female': 0
})

In [120]:
new_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7043 entries, 0 to 7042
Data columns (total 25 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   City               7043 non-null   object 
 1   Zip Code           7043 non-null   int64  
 2   Latitude           7043 non-null   float64
 3   Longitude          7043 non-null   float64
 4   Gender             7043 non-null   int64  
 5   Senior Citizen     7043 non-null   int64  
 6   Partner            7043 non-null   int64  
 7   Dependents         7043 non-null   int64  
 8   Tenure Months      7043 non-null   int64  
 9   Phone Service      7043 non-null   int64  
 10  Multiple Lines     7043 non-null   int64  
 11  Internet Service   7043 non-null   object 
 12  Online Security    7043 non-null   int64  
 13  Online Backup      7043 non-null   int64  
 14  Device Protection  7043 non-null   int64  
 15  Tech Support       7043 non-null   int64  
 16  Streaming TV       7043 

We have successfully replace all Yes/No categoricals with 1/0 values